File: SAUCEDEMO-Playwright/Python/utils/driver_factory.py
from playwright.sync_api import Page, Browser, BrowserContext, expect, Playwright, sync_playwright


class WebDriverManagement:
    def __init__(self, config, test_data):
        self.pages = {}
        self.config = config
        self.test_data = test_data
        self.browser : Browser = None
        self.context : BrowserContext = None
        self.page : Page = None
        self.Playwright: Playwright = None
        self.expect = expect
        
        
    def start_browser(self):
        ''' create a customable and manageable web driver '''
        try :
            print('Browser initialize.')
            self.Playwright = sync_playwright().start()
            web_driver = self.Playwright.chromium
            self.browser = web_driver.launch(
                headless=self.config['HEADLESS'],
                args=['--start-maximized'],
                slow_mo=self.config.get('INTERACT_DELAY',500)
            )
            self.context = self.browser.new_context(no_viewport=True)
            global_wait = self.config.get('GLOBAL_WAIT', 10000)
            navigation_wait = self.config.get('NAV_WAIT', 10000)
            self.context.set_default_navigation_timeout(navigation_wait)
            self.context.set_default_timeout(global_wait)
            print('Complete init browser.')
            return self.context
        except Exception as e:
            raise e    
    def close_browser(self):
        if self.context:
            self.context.close()
            self.browser.close()
            self.Playwright.stop()
        else: 
            print("No Browser is opened.")
            
    def _ensure_browser_is_running(self):
        if not self.browser:
            self.start_browser()
        elif not self.context:
            self.context = self.browser.new_context(no_viewport=True)
            
    def create_page(self,url,page_name):
        self._ensure_browser_is_running()
        new_page = self.context.new_page()
        new_page.goto(url, timeout=10000, wait_until="load")
        
        self.page = new_page
        self.pages[page_name] = new_page
        
        return new_page
    
    def switch_to_page(self, page_name):
        if page_name in self.pages:
            self.page = self.pages[page_name]
            self.page.bring_to_front()
        else:
            raise Exception(f"There is no such page name {page_name} stored in pages.")

File: SAUCEDEMO-Playwright/TypeScript/utils/driverFactory.ts
import {
  Page,
  expect,
  Browser,
  BrowserContext,
  chromium,
} from "@playwright/test";
import * as configFile from "../config/config";
import * as testData from "../config/testdata";

export class WebDriverManagement {
  public pages: Record<string, Page> = {};
  public config = configFile;
  public testData = testData;
  public expect = expect;
  public page!: Page;
  public browser!: Browser;
  public context!: BrowserContext;

  async startBrowser() {
    this.browser = await chromium.launch({
      slowMo: 500,
      args: ["--start-maximized"],
    });
    this.context = await this.browser.newContext({ viewport: null });
    this.context.setDefaultTimeout(this.config.globalWait);
    this.context.setDefaultNavigationTimeout(this.config.navWait);
    console.log(`complete init browser.`);
  }

  async closeBrowser() {
    if (this.context) {
      await this.context.close();
      await this.browser.close();
    } else if (this.browser) {
      await this.browser.close();
    } else console.log(`No context or browser had opened.`);
  }

  async closeContext() {
    if (this.context) await this.context.close();
  }

  async ensureBrowserIsRunning() {
    if (!this.browser) await this.startBrowser();
    else if (!this.context)
      this.context = await this.browser.newContext({ viewport: null });
  }

}
